activerecord/lib/active_record/attribute_methods/primary_key.rb

```diff
@@ -116,7 +114,7 @@ def get_primary_key(base_name) #:nodoc:
           #
           #   Project.primary_key # => "foo_id"
           def primary_key=(value)
-            @primary_key        = value && value.to_s
+            @primary_key        = value && -value.to_s
```

Comment:
What does the minus do here?

Reply:
Ruby 2.5 and higher has a string deduplication (fstring table), `String#-@` registers the string to fstring table and fetches the deduplicated instance.
The most `@primary_key`s are `"id"`, the deduplication will reduce memory usage.